Use CSISnapshotTimeout to set the snapshot timeout

dzaninovic opened the pull request:
Here is a fix for the issue where CSISnapshotTimeout is not getting used by the plugin as reported here:
https://github.com/vmware-tanzu/velero/issues/6261

I ran these tests:
1. Edit snapshot-controller deployment to have 0 replicas so it would fail to create the snapshot.
Run a backup with CSISnapshotTimeout set to 1 minute and confirm from the log that timeout is honored.
2. Same test as above but without setting the CSISnapshotTimeout.  Timeout was 10 minutes as expected.
3. Let snapshot-controller deployment have 2 replicas and confirm that backup worked as normal.
4. Run "make test" and confirm that all tests passed.

blackpiglet commented:
Please check the comments here
https://github.com/vmware-tanzu/velero/issues/6261#issuecomment-1549539789
Suggest putting this PR on hold until an agreement is reached.

anshulahuja98 commented:
@dzaninovic you need to signoff your commit for DCO to succeed
you can do so by doing git commit --signoff

We have overall reached a conclusion on the tracking issue and hopefully this PR should go in

anshulahuja98 commented on internal/util/util.go:
```diff
@@ -158,9 +158,11 @@ func GetVolumeSnapshotContentForVolumeSnapshot(volSnap *snapshotv1api.VolumeSnap
-	// TODO: make this timeout configurable.
+	// We'll wait 10m for the VSC to be reconciled polling every 5s unless csiSnapshotTimeout is set
 	timeout := 10 * time.Minute
+	if csiSnapshotTimeout > 0 {
```
should we do null checks before this?

anshulahuja98 replied:
https://github.com/vmware-tanzu/velero/blob/25fb08b3c27955cfb04ab72037faa996a117c8a7/pkg/controller/backup_controller.go#LL330C2-L333C3

You might want to add default case handling similar to the core code polling.

dzaninovic commented:
@anshulahuja98,

Here are the answers to your review comments that I got in my email:

1. We don't need checks for nil as there is nothing that can be nil except *backup which is already used in the caller of GetVolumeSnapshotContentForVolumeSnapshot() few lines above my code.

2. Velero repository has defaultCSISnapshotTimeout defined in pkg/cmd/server/server.go.  I can't use it directly or indirectly unless I modify Velero repository also so I just added defaultCSISnapshotTimeout const to the same source file where I am using it.

3. I added a signoff to my commit messages.

draghuram commented:
We discussed this PR in today's community call. @eemcmullan, please take a look. In general, there is a consensus that the PR can be merged.
